### ISSUE TITLE
Bundling of the Discord Game SDK violates the GPL

### ISSUE THREAD

**Issue by dhrdlicka** [86Box/86Box#1606]
**Describe the bug**
The Discord module interacts with the Game SDK in such way that they are considered a single combined project by the GPL, which means that the Game SDK must be licensed and distributed under the terms of the GPL.

**Additional context**
From the [Frequently Asked Questions about the GNU Licenses](http://www.gnu.org/licenses/gpl-faq.en.html):

> If the main program dynamically links plug-ins, and they make function calls to each other and share data structures, we believe they form a single combined program, which must be treated as an extension of both the main program and the plug-ins.

> If the main program and the plugins are a single combined program then this means you must license the plug-in under the GPL or a GPL-compatible free software license and distribute it with source code in a GPL-compliant way.

**Comment by Calinou** [comment]
Using the deprecated-but-still-working discordrpc library is an alternative, as it's MIT-licensed.

**Comment by nikitalita** [comment]
I wrote up an explanation [here](https://gist.github.com/nikitalita/b7fe0ab1288788423529689546fd62ee), but I'll reproduce it here:

(Disclaimer: IANAL, this does not constitute legal advice, seek a lawyer for more guidance if necessary)

GPL and other FOSS licenses are only concerned with **distribution**; indeed, restrictions on use are forbidden by free software and open source definitions, it can only enforce restrictions on distribution. So, GPL clauses come into force whenever a GPL program is distributed to other people. 

The thing that's at issue here is the FSF's interpretation of copyright law when it comes to software and derivative works. 
* First, the FSF takes the position that a program that dynamic links against a library at runtime with constitutes a derivative work; you've combined a program and a library to make a new program. 
* Second, "collectivity": They take the position that when you distribute said program with said library, you are distributing a "combined" program that constitutes a derivative work, **even if** said combination actually happens at runtime. The argument is that when you distribute them together, they are not "independent and seperate works", respectively. 
* (There's also the "interdependency" argument, but that doesn't apply here since it's an optional feature.)

So, according to this interpretation, if you dynamically link a GPL program with a library, the GPL would apply to both the libraries and the program itself. They can't enforce this against programs that link against DLLs that are not distributed with the program, since that's a restriction on use (which is why they have that Systems libraries exception in the first place). But, they **can** enforce that against programs that are **distributed with** those DLLs.

Nevertheless, their ability to enforce that rests on their interpretation of copyright law to actually be correct, and neither their interpretation of what constitutes derivative works or their interpretation of "collectivity" have actually been tested in court. Several copyright experts take the opinion that it is not correct.

However, most companies thus far have acted like this is the case and generally avoid doing this. Indeed, Discord seems to take this position themselves according to this from their [Store Distribution Agreement](https://github.com/discord/discord-api-docs/blob/master/docs/Store_Distribution_Agreement.md):

> 2.6 Open Source Software. Except as otherwise expressly agreed in writing, Developer shall not distribute via Discord Store, or combine any Discord materials with, open source or other software that is licensed under terms that purport to bind Discord to contractual obligations (e.g., the GNU General Public License or Lesser General Public License).

## TL;DR:
So, if you want to include Discord game SDK integration with a GPL program, the option that would cause the least friction while still being safe would likely be to prompt the user to automatically download and install it when they enable it for the first time.

Further reading:
[Software Interactions and the GNU General
Public License](https://www.jolts.world/index.php/jolts/article/download/44/74)
[GPL - the Linking Debate](https://web.archive.org/web/20130514013222/https://moorcrofts.com/documents/GPL%20-%20the%20Linking%20Debate.pdf)

**Comment by waltje** [comment]
Hi:
> (Disclaimer: IANAL, this does not constitute legal advice, seek a lawyer for more guidance if necessary)Neither am I, but I actually DO have a good one ;-)
> GPL and other FOSS licenses are only concerned with distribution...

Correct.
------------------
The thing that's at issue here is the FSF's interpretation of copyright law when it comes to software and derivative works.   
   - First, the FSF takes the position that a program that dynamic links against a library at runtime with constitutes a derivative work; you've combined a program and a library to make a new program.
   - Second, "collectivity": They take the position that when you distribute said program with said library, you are distributing a "combined" program that constitutes a derivative work, even if said combination actually happens at runtime. The argument is that when you distribute them together, they are not "independent and seperate works", respectively.
   - (There's also the "interdependency" argument, but that doesn't apply here since it's an optional feature.)   


------------------------

> So, according to this interpretation, if you dynamically link a GPL program with a library,> the GPL would apply to both the libraries and the program itself.Yes, and this has been under heavy debate since the origins of time.
> They can't enforce this against programs that link against DLLs that are not distributed
.. because of this.
> But, they can enforce that against programs that are distributed with those DLLs.
Not according to MY laywer, and he has been doing IP law for 30 years.  If I create aGPL program, which *can* be used with ".. other modules" (dlls, feature plugins, whatever)then those additional modules are NOT AN INTEGRAL PART of said GPL program.
The grey area here is, if those ".. modules" are hard-linked to said GPL program, so, linkedfrom a static library, or hard-linked against such a module (.dll or .so.)  In those cases, theGPL program cannot function without those modules, and, so, DOES form an integral partof it.
> Nevertheless, their ability to enforce that rests on their interpretation of copyright law to actually be correct, and neither their interpretation of what constitutes derivative works or their interpretation of "collectivity" have actually been tested in court. 
My lawyer agrees with you (as do I.)

> Several copyright experts take the opinion that it is not correct.See above.

The simplest solution is to, indeed, ignore the issue altogether.  If you're a little skiddish 
of suddenly being yanked out of bed by the feds because of that, well, do not hardlink suchmodules, dynamic-link them.  I added that to 86Box for a reason, you know.
--Fred (waltje).

**Comment by OBattler** [comment]
We do dynamically link the Discord Game SDK, so 86Box can function without it. As such, I do not see how they can be thought of a single, combined program.

**Comment by OBattler** [comment]
I'm going to close this as being settled with a decision to keep the status quo.
